perception/autoware_traffic_light_selector/src/traffic_light_selector_node.cpp

```diff
@@ -63,89 +58,78 @@ void TrafficLightSelectorNode::objectsCallback(
     const auto & rough_roi = rough_rois_map[traffic_light_id];
+    const auto & expect_roi = expected_rois_msg_rois.roi;
 
-    for (const auto & detected_roi : det_rois) {
-      // check if the detected roi is inside the rough roi
+    for (const auto & detected_roi : detected_rois) {
       if (!utils::isInsideRoughRoi(detected_roi, rough_roi)) {
         continue;
       }
-      int dx = static_cast<int>(detected_roi.x_offset) - static_cast<int>(expect_roi.x_offset);
-      int dy = static_cast<int>(detected_roi.y_offset) - static_cast<int>(expect_roi.y_offset);
-      cv::Mat det_roi_shifted = utils::shiftAndPaddingImage(det_roi_img, dx, dy);
-      double iou = utils::getIoUOf2BinaryImages(expect_roi_img, det_roi_shifted);
-      if (iou > max_matching_score) {
-        max_matching_score = iou;
-        det_roi_shift_x = dx;
-        det_roi_shift_y = dy;
+
+      // shift expect roi because the location of detected roi is better that it
+      int32_t shift_x, shift_y;
+      utils::computeCenterOffset(detected_roi, expect_roi, shift_x, shift_y);
+
+      std::map<uint8_t, RegionOfInterest> expect_rois_shifted_map;
+      for (const auto & expected_rois_msg_rois : expected_rois_msg->rois) {
+        const auto expect_roi_shifted = utils::getShiftedRoi(
+          expected_rois_msg_rois.roi, image_width, image_height, shift_x, shift_y);
+        expect_rois_shifted_map[expected_rois_msg_rois.traffic_light_id] = expect_roi_shifted;
       }
-    }
-  }
 
-  for (const auto & expect_roi : expected_rois_msg->rois) {
-    // check max IOU after shift
-    double max_iou = -1.0;
-    sensor_msgs::msg::RegionOfInterest max_iou_roi;
-    for (const auto & detected_roi : detected_traffic_light_msg->feature_objects) {
-      // shift detected roi by det_roi_shift_x, det_roi_shift_y and calculate IOU
-      sensor_msgs::msg::RegionOfInterest detected_roi_shifted = detected_roi.feature.roi;
-      // fit top lef corner of detected roi to inside of image
-      detected_roi_shifted.x_offset = std::clamp(
-        static_cast<int>(detected_roi.feature.roi.x_offset) - det_roi_shift_x, 0,
-        static_cast<int>(image_width - detected_roi.feature.roi.width));
-      detected_roi_shifted.y_offset = std::clamp(
-        static_cast<int>(detected_roi.feature.roi.y_offset) - det_roi_shift_y, 0,
-        static_cast<int>(image_height - detected_roi.feature.roi.height));
-
-      double iou = utils::getGenIoU(expect_roi.roi, detected_roi_shifted);
-      if (iou > max_iou) {
-        max_iou = iou;
-        max_iou_roi = detected_roi.feature.roi;
+      // check total IoU after all expect roi shift
+      double total_max_iou = 0.0;
+      std::map<uint8_t, RegionOfInterest> total_max_iou_rois_map;
+      evaluateWholeRois(
+        detected_rois, expect_rois_shifted_map, total_max_iou, total_max_iou_rois_map);
+
+      if (total_max_iou > final_iou) {
+        final_iou = total_max_iou;
+        final_rois_map = total_max_iou_rois_map;
```

Comment:
Is better to use `std::move` ?
```suggestion
        final_rois_map = std::move(total_max_iou_rois_map);
```